html_tag_collector/collector.py

```diff
@@ -231,70 +237,181 @@ def parse_response(url_response):
     """Parses relevant HTML tags from a Response object into a dictionary.
 
     Args:
-        url_response (list[dict]): List of dictionaries containing urls and theeir responses.
+        url_response (list[dict]): List of dictionaries containing urls and their responses.
 
     Returns:
-        list[dict]: List of dictionaries containing urls and relevant HTML tags.
+        dict: Dictionary containing the url and relevant HTML tags.
     """
-    remove_excess_whitespace = lambda s: " ".join(s.split()).strip()
-    
-    tags = {}
+    tags = Tags()
     res = url_response["response"]
-    tags["index"] = url_response["index"]
+    tags.index = url_response["index"]
 
-    # Drop hostname from urls to reduce training bias
+    tags.url, tags.url_path = get_url(url_response)
+
+    tags.root_page_title = remove_excess_whitespace(root_url_cache.get_title(tags.url))
+
+    verified, tags.http_response = verify_response(res)
+    if verified is False:
+        return asdict(tags)
+
+    parser = get_parser(res)
+    if parser is False:
+        return asdict(tags)
+
+    try:
+        soup = BeautifulSoup(res.html.html, parser)
+    except (bs4.builder.ParserRejectedMarkup, AssertionError, AttributeError):
+        return asdict(tags)
+
+    tags.html_title = get_html_title(soup)
+
+    tags.meta_description = get_meta_description(soup)
+
+    tags = get_header_tags(tags, soup)
+
+    tags.div_text = get_div_text(soup)
+
+    # Prevents most bs4 memory leaks
+    if soup.html:
+        soup.html.decompose()
+
+    return asdict(tags)
+
+
+def get_url(url_response):
+    """Returns the url and url_path.
+
+    Args:
+        url_response (list[dict]): List of dictionaries containing urls and their responses.
+
+    Returns:
+        (str, str): Tuple with the url and url_path.
+    """
     url = url_response["url"][0]
-    tags["url"] = url
-    if not url.startswith("http"):
-        url = "https://" + url
-    tags["url_path"] = urlparse(url).path[1:]
+    new_url = url
+    if not new_url.startswith("http"):
+        new_url = "https://" + new_url
+
+    # Drop hostname from urls to reduce training bias
+    url_path = urlparse(new_url).path[1:]
+    # Remove trailing backslash
+    if url_path and url_path[-1] == "/":
+        url_path = url_path[:-1]
 
-    tags["html_title"] = ""
-    tags["meta_description"] = ""
-    tags["root_page_title"] = remove_excess_whitespace(root_url_cache.get_title(tags["url"]))
+    return url, url_path
 
+
+def verify_response(res):
+    """Verifies the webpage response is readable and ok.
+
+    Args:
+        res (HTMLResponse|Response): Response object to verify.
+
+    Returns:
+        (bool, int): A tuple containing False if verification fails, True otherwise and the http response code
+    """
+    # The response is None if there was an error during connection, meaning there is no content to read
     if res is None:
-        tags["http_response"] = -1
-        return tags
+        return False, -1
```

Comment:
NIT: I won't block on this, but recommend converting multi-value responses to named tuples. At a glance, I don't know what each of these values represent, and a named tuple will greatly clarify this.

Reply:
@maxachis like this? I've never used a namedtuple before so correct me if this isn't what you had in mind

```python
def verify_response(res):
    """Verifies the webpage response is readable and ok.

    Args:
        res (HTMLResponse|Response): Response object to verify.

    Returns:
        VerifiedResponse(bool, int): A named tuple containing False if verification fails, True otherwise and the http response code.
    """
    VerifiedResponse = namedtuple("VerifiedResponse", "verified http_response")
    # The response is None if there was an error during connection, meaning there is no content to read
    if res is None:
        return VerifiedResponse(False, -1)

    # If the connection did not return a 200 code, we can assume there is no relevant content to read
    http_response = res.status_code
    if not res.ok:
        return VerifiedResponse(False, http_response)
        
    return VerifiedResponse(True, http_response)
```

Reply:
Yep! Just make sure to have it outside of the function so that other parts of the functions can reference it (and also don't forget that the fields are a comma-delimited list, so  VerifiedResponse = namedtuple("VerifiedResponse", ["verified", "http_response"])